src/map_field.cpp

```diff
@@ -1707,25 +1694,44 @@ void map::creature_in_field( Creature &critter )
             continue;
         }
         const field_type_id cur_field_id = cur_field_entry.get_field_type();
-        const effect field_fx = cur_field_entry.field_effect();
-        if( field_fx.is_null() ||
-            critter.is_immune_field( cur_field_id ) || field_fx.get_id().is_empty() ||
-            critter.is_immune_effect( field_fx.get_id() ) ) {
-            continue;
-        }
-        player *u = critter.as_player();
-        if( u && cur_field_entry.inside_immune() ) {
-            // If we are in a vehicle figure out if we are inside (reduces effects usually)
-            // and what part of the vehicle we need to deal with.
-            if( u->in_vehicle ) {
-                if( const optional_vpart_position vp = veh_at( u->pos() ) ) {
-                    if( vp->is_inside() ) {
-                        continue;
-                    }
-                }
+
+        for( const auto &fe : cur_field_entry.field_effects() ) {
+            if( in_vehicle && fe.immune_in_vehicle ) {
+                continue;
+            }
+            if( inside_vehicle && fe.immune_inside_vehicle ) {
+                continue;
+            }
+            if( !inside_vehicle && fe.immune_outside_vehicle ) {
+                continue;
+            }
+            if( in_vehicle && !one_in( fe.chance_in_vehicle ) ) {
+                continue;
+            }
+            if( inside_vehicle && !one_in( fe.chance_inside_vehicle ) ) {
+                continue;
+            }
+            if( !inside_vehicle && !one_in( fe.chance_outside_vehicle ) ) {
+                continue;
+            }
+
+            const effect field_fx = fe.get_effect();
+            if( field_fx.is_null() ||
+                critter.is_immune_field( cur_field_id ) || field_fx.get_id().is_empty() ||
```

Comment:
Now that `field_entry::field_effects` returns a vector, it can return an empty vector to signal "no effect".

There: why allow a null effect object to be returned? Such a null object (or an object with an empty id) hints that the implementation of `field_entry::field_effects` is wrong or that the JSON data is wrong.

It should be assumed that `field_entry::field_effects` only returns valid effects (and therefor simply don't check the id at all).

This is all that should be needed here:
```suggestion
            if( critter.is_immune_field( cur_field_id ) || critter.is_immune_effect( field_fx.get_id() ) ) {
```